README.md
# flutter_query_builder

⬛️ - Whether you're a beginner or an experienced developer, flutter_query_builder simplifies database interactions, allowing you to create, retrieve, update, and delete records with ease. Say goodbye to manual SQL queries and embrace the simplicity and power of flutter_query_builder for all your database operations .

⬛️ - flutter_query_builder empowers Flutter developers to work with SQL databases more efficiently, reducing development time and effort.  developers can focus on building robust applications without getting lost in intricate SQL syntax .

⬛️ - It should be noted that this package is based on the sqflite package 😁 . 
## Get started

### Add dependency

```yaml
dependencies:
  flutter_query_builder: ^1.0.0
```

⬛️ - import sqflite and start initializing your own database.

```dart
import 'dart:async';
import 'package:sqflite/sqflite.dart';

class DataBaseManagment {
  static late Database database;

  static Future<Database> initDatabase({
    required String dbName,
    int? version,
    FutureOr<void> Function(Database)? onOpen,
    FutureOr<void> Function(Database, int)? onCreate,
    FutureOr<void> Function(Database)? onConfigure,
    FutureOr<void> Function(Database, int, int)? onDowngrade,
    FutureOr<void> Function(Database, int, int)? onUpgrade,
    bool? readOnly = false,
    bool? singleInstance = true,
  }) async {
    var databasesPath = await getDatabasesPath();
    String path = '$databasesPath/$dbName';
    // open the database
    try {
      database = await openDatabase(
        path,
        version: version,
        onOpen: onOpen,
        onConfigure: onConfigure,
        onDowngrade: onDowngrade,
        onUpgrade: onUpgrade,
        singleInstance: singleInstance,
        readOnly: readOnly,
        onCreate: onCreate,
      );
      print('$dbName  DataBase initialized succeffuly with path $path');
    } catch (e) {
      print(
          'Something went wrong when initializing $dbName  DataBase / error : $e');
    }
    return database;
  }
}

void main() async {
  await DataBaseManagment.initDatabase(dbName: 'test', version: 1);
  runApp(const MyApp());
}

```
### first thing first ... we assume that we have this dummy data to explain main functionallity

```dart
TableModel student = TableModel(
  name: 'Student',
  columns: [
    ColumnModel(
      name: 'studentId',
      isAutoincrement: true,
      isPrimaryKey: true,
      isNotNull: true,
      isInteger: true,
    ),
    ColumnModel(name: 'name', isText: true),
    ColumnModel(name: 'collageId', isInteger: true),
  ],
);

TableModel collage = TableModel(
  name: 'Collage',
  columns: [
    ColumnModel(
      name: 'collageId',
      isAutoincrement: true,
      isPrimaryKey: true,
      isNotNull: true,
      isInteger: true,
    ),
    ColumnModel(name: 'name', isVarchar: true,varcharCharCount: 50),
    ColumnModel(name: 'address', isText: true),
  ],
);
                         
```

### Super easy to use

⬜️ - creating a table :

```dart

await QueryBuilder.createTable(
        table: collage,
        whenError: (error) {
            log('Error 👾: $error');
        },
    );
                              
```

⬜️ - get table content :

```dart


await QueryBuilder.getTableContent(
        tableName: collage.name,
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                
                              
```

⬜️ - rename a table :

```dart

await QueryBuilder.renameTable(
        table: student,
        newName: 'first Table',
        whenError: (error) {
            log('Error 👾: $error');
        },
);              
                              
```

⬜️ - get tables Names :

```dart


await QueryBuilder.getTablesNames(
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                
                              
```

⬜️ - delete a table :

```dart

await QueryBuilder.dropTable(
        table: student,
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                          
                              
```

⬜️ - add new column :

```dart


await QueryBuilder.addNewColumn(
        table: student,
        column: ColumnModel(
            name: 'bio',
            isReal: true,
        ),
        whenError: (error) {
                log('Error 👾: $error');
        },
);
                
                              
```

⬜️ - update column value :

```dart

await QueryBuilder.updateColumnValue(
        table: student,
        columnName: 'ID_adress',
        newColumnValue: 'NewName',
        condition: ConditionModel(
                key: 'id',
                condition: ConditionType.equalTo,
                val: 5,
        ),
        whenError: (error) {},
);
                
```

⬜️ - get all names of table columns :

```dart         

await QueryBuilder.getColumnNames(student.name);

```

⬜️ - rename a column :

```dart


await QueryBuilder.renameColumn(
        table: student,
        oldName: 'name',
        newName: 'newName',
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                

```

⬜️ - insert a record :

```dart

await QueryBuilder.insertRecord(
        table: collage,
        record: RecordModel(
                    data: [
                      RecordItemModel(columnName: 'name', value: 'Emran Maher Al-daqaq'),
                      RecordItemModel(columnName: 'address', 'Duff Fork Villas'),
                      RecordItemModel(columnName: 'collageId', value: 5),
                    ]),
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                                       
```

⬜️ - update a record :

```dart

                  await QueryBuilder.updateRecord(
                    table: student,
                    newRecord: [
                      RecordItemModel(columnName: 'name', value: 'new name'),
                      RecordItemModel(columnName: 'age', value: '10'),
                      RecordItemModel(columnName: 'ID_adress', value: 65),
                      RecordItemModel(columnName: 'have', value: false),
                      RecordItemModel(columnName: 'haveCat', value: 0),
                    ],
                    condition: ConditionModel(
                        key: 'id', condition: ConditionType.notEqualTo, val: 6),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                                              
```

⬜️ - delete a record :

```dart


await QueryBuilder.deleteRecord(
        table: student,
        condition: ConditionModel(
                        key: 'haveTV',
                        condition: ConditionType.equalTo,
                        val: 0,
                    ),
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                                
                              
```

🔲 - Union & UnionAll Queries :

```dart

await QueryBuilder.union(
        type: UnionType.union, // ~ or you can use UnionType.unionAll
        firstRowSelect: 'SELECT bio FROM ${student.name}',
        secondRowSelect: 'SELECT bio FROM ${student.name}',
        orderByColumn: 'bio',
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                            
                              
```

🔲 - join Queries :

```dart
/*
join types: 
JoinType.inner;
JoinType.left;
 ~~ RIGHT and FULL OUTER JOINs are not currently supported
*/
await QueryBuilder.join(
        type: JoinType.inner,
        rowSelect:
                    'SELECT ${collage.name}.address,${student.name}.name FROM ${student.name}',
        secondTableName: collage.name,
        firstColumnName: '${collage.name}.collageId',
        secondColumnName: '${student.name}.collageId',
        whenError: (error) {
            log('Error 👾: $error');
        },
);
                           
                              
```

⬜️ - Row Query :

```dart

await QueryBuilder.rawQuery(
        'query',
        whenError: (error) {
            log('Error 👾: $error');
        },
);                         
                              
```

⬜️ - deleta the data base :

```dart

await deleteDataBase(String dbName);                     
                              
```

### super easy ... wasn't it?

## ❤️ Found this project useful?

If you found this useful, then please consider giving it a ⭐ on Github and sharing it with your friends via social media.

## ❤️ Contact Info
## 🔲  LinkedIn : https://www.linkedin.com/in/emran-al-daqaq
## 🔲  Facebook : https://www.facebook.com/emran.aldakak?mibextid=ZbWKwL


CHANGELOG.md
## 1.0.0

- adding main functionallity

## 1.0.1

- example added successfully

## 1.0.2

- real example added successfully



example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_query_builder_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/flutter_query_builder_test.dart
import 'package:flutter_query_builder/flutter_query_builder.dart';

void main() {}


example/lib/home.dart
import 'package:flutter/material.dart';
import 'dart:developer';
import 'package:flutter_query_builder/flutter_query_builder.dart';

import 'common/dummy_data.dart';

class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        body: Padding(
          padding: const EdgeInsets.all(8.0),
          child: ListView(
            children: [
              // ~ Initializing data base
              const Center(child: Text('Initializing data base')),
              ButtonsWidget(
                title: 'Initializing data base',
                onTap: () async {
                  await DataBaseManagment.initDatabase(
                      dbName: 'test1', version: 1);
                },
              ),

              // ~ Tables
              const Center(child: Text('Tables')),
              ButtonsWidget(
                title: 'Create Table',
                onTap: () async {
                  await QueryBuilder.createTable(
                    table: collage,
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Get Table Content',
                onTap: () async {
                  await QueryBuilder.getTableContent(
                    tableName: collage.name,
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Rename Table',
                onTap: () async {
                  await QueryBuilder.renameTable(
                    table: student,
                    newName: 'first Table',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Get Table Names',
                onTap: () async {
                  await QueryBuilder.getTablesNames(
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Drop Table',
                onTap: () async {
                  await QueryBuilder.dropTable(
                    table: student,
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              // ~ Columns
              const Center(child: Text('Columns')),

              ButtonsWidget(
                title: 'Add New Column',
                onTap: () async {
                  await QueryBuilder.addNewColumn(
                    table: student,
                    column: ColumnModel(
                      name: 'bio',
                      isReal: true,
                    ),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              ButtonsWidget(
                title: 'Update Column Value',
                onTap: () async {
                  await QueryBuilder.updateColumnValue(
                    table: student,
                    columnName: 'ID_adress',
                    newColumnValue: 'hi',
                    condition: ConditionModel(
                      key: 'id',
                      condition: ConditionType.equalTo,
                      val: 5,
                    ),
                    whenError: (error) {},
                  );
                },
              ),

              ButtonsWidget(
                title: 'Column Names In Table',
                onTap: () async {
                  await QueryBuilder.getColumnNames(collage.name);
                },
              ),
              ButtonsWidget(
                title: 'Rename Column',
                onTap: () async {
                  await QueryBuilder.renameColumn(
                    table: student,
                    oldName: 'have4324',
                    newName: 'havek',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              // ~ Reords

              const Center(child: Text('Records')),

              ButtonsWidget(
                title: 'Insert Record',
                onTap: () async {
                  await QueryBuilder.insertRecord(
                    table: collage,
                    record: RecordModel(data: [
                      RecordItemModel(
                          columnName: 'name', value: 'Duff Fork Villas'),
                      RecordItemModel(columnName: 'address', value: 'test2'),
                      RecordItemModel(columnName: 'collageId', value: 5),
                    ]),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Update Record',
                onTap: () async {
                  await QueryBuilder.updateRecord(
                    table: student,
                    newRecord: [
                      RecordItemModel(columnName: 'name', value: 'Erhamny'),
                      RecordItemModel(columnName: 'age', value: '10'),
                      RecordItemModel(columnName: 'ID_adress', value: 65),
                      RecordItemModel(columnName: 'have', value: false),
                      RecordItemModel(columnName: 'haveCat', value: 0),
                    ],
                    condition: ConditionModel(
                        key: 'id', condition: ConditionType.notEqualTo, val: 6),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Delete Record',
                onTap: () async {
                  await QueryBuilder.deleteRecord(
                    table: student,
                    condition: ConditionModel(
                        key: 'haveTV',
                        condition: ConditionType.equalTo,
                        val: 0),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              // ~ Union

              const Center(child: Text('Union')),

              ButtonsWidget(
                title: 'Union',
                onTap: () async {
                  await QueryBuilder.union(
                    type: UnionType.union,
                    firstRowSelect: 'SELECT bio FROM ${student.name}',
                    secondRowSelect: 'SELECT bio FROM ${student.name}',
                    orderByColumn: 'bio',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Union All',
                onTap: () async {
                  await QueryBuilder.union(
                    type: UnionType.unionAll,
                    firstRowSelect: 'SELECT bio FROM ${student.name}',
                    secondRowSelect: 'SELECT bio FROM ${student.name}',
                    orderByColumn: 'bio',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              // ~ Inner Join
              const Center(child: Text('Join')),
              ButtonsWidget(
                title: 'Inner Join',
                onTap: () async {
                  JoinType.inner;
                  JoinType.left;
                  await QueryBuilder.join(
                    type: JoinType.inner,
                    rowSelect:
                        'SELECT ${collage.name}.address,${student.name}.name FROM ${student.name}',
                    secondTableName: collage.name,
                    firstColumnName: '${collage.name}.collageId',
                    secondColumnName: '${student.name}.collageId',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              // ~ left Join
              ButtonsWidget(
                title: 'Left Join',
                onTap: () async {
                  await QueryBuilder.join(
                    type: JoinType.left,
                    rowSelect:
                        'SELECT ${collage.name}.address,${student.name}.name FROM ${student.name}',
                    secondTableName: collage.name,
                    firstColumnName: '${collage.name}.collageId',
                    secondColumnName: '${student.name}.collageId',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              // ~ Row Query

              const Center(child: Text('Row Query')),

              ButtonsWidget(
                title: 'Row Query',
                onTap: () async {
                  await QueryBuilder.rawQuery(
                    'query',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class ButtonsWidget extends StatelessWidget {
  const ButtonsWidget({
    super.key,
    required this.onTap,
    required this.title,
  });
  final VoidCallback onTap;
  final String title;
  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: ElevatedButton(
        onPressed: onTap,
        child: Text(title),
      ),
    );
  }
}


example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "flutter_query_builder_example");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "flutter_query_builder_example");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_query_builder/flutter_query_builder.dart';
import 'dart:developer';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'flutter_query_builder',
      home: HomeScreen(),
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        body: Padding(
          padding: const EdgeInsets.all(8.0),
          child: ListView(
            children: [
              // ~ Initializing data base
              const Center(child: Text('Initializing data base')),
              ButtonsWidget(
                title: 'Initializing data base',
                onTap: () async {
                  await DataBaseManagment.initDatabase(
                      dbName: 'test1', version: 1);
                },
              ),

              // ~ Tables
              const Center(child: Text('Tables')),
              ButtonsWidget(
                title: 'Create Table',
                onTap: () async {
                  await QueryBuilder.createTable(
                    table: collage,
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Get Table Content',
                onTap: () async {
                  await QueryBuilder.getTableContent(
                    tableName: collage.name,
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Rename Table',
                onTap: () async {
                  await QueryBuilder.renameTable(
                    table: student,
                    newName: 'first Table',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Get Table Names',
                onTap: () async {
                  await QueryBuilder.getTablesNames(
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Drop Table',
                onTap: () async {
                  await QueryBuilder.dropTable(
                    table: student,
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              // ~ Columns
              const Center(child: Text('Columns')),

              ButtonsWidget(
                title: 'Add New Column',
                onTap: () async {
                  await QueryBuilder.addNewColumn(
                    table: student,
                    column: ColumnModel(
                      name: 'bio',
                      isReal: true,
                    ),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              ButtonsWidget(
                title: 'Update Column Value',
                onTap: () async {
                  await QueryBuilder.updateColumnValue(
                    table: student,
                    columnName: 'ID_adress',
                    newColumnValue: 'hi',
                    condition: ConditionModel(
                      key: 'id',
                      condition: ConditionType.equalTo,
                      val: 5,
                    ),
                    whenError: (error) {},
                  );
                },
              ),

              ButtonsWidget(
                title: 'Column Names In Table',
                onTap: () async {
                  await QueryBuilder.getColumnNames(collage.name);
                },
              ),
              ButtonsWidget(
                title: 'Rename Column',
                onTap: () async {
                  await QueryBuilder.renameColumn(
                    table: student,
                    oldName: 'have4324',
                    newName: 'havek',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              // ~ Reords

              const Center(child: Text('Records')),

              ButtonsWidget(
                title: 'Insert Record',
                onTap: () async {
                  await QueryBuilder.insertRecord(
                    table: collage,
                    record: RecordModel(data: [
                      RecordItemModel(
                          columnName: 'name', value: 'Duff Fork Villas'),
                      RecordItemModel(columnName: 'address', value: 'test2'),
                      RecordItemModel(columnName: 'collageId', value: 5),
                    ]),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Update Record',
                onTap: () async {
                  await QueryBuilder.updateRecord(
                    table: student,
                    newRecord: [
                      RecordItemModel(columnName: 'name', value: 'Erhamny'),
                      RecordItemModel(columnName: 'age', value: '10'),
                      RecordItemModel(columnName: 'ID_adress', value: 65),
                      RecordItemModel(columnName: 'have', value: false),
                      RecordItemModel(columnName: 'haveCat', value: 0),
                    ],
                    condition: ConditionModel(
                        key: 'id', condition: ConditionType.notEqualTo, val: 6),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Delete Record',
                onTap: () async {
                  await QueryBuilder.deleteRecord(
                    table: student,
                    condition: ConditionModel(
                        key: 'haveTV',
                        condition: ConditionType.equalTo,
                        val: 0),
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),

              // ~ Union

              const Center(child: Text('Union')),

              ButtonsWidget(
                title: 'Union',
                onTap: () async {
                  await QueryBuilder.union(
                    type: UnionType.union,
                    firstRowSelect: 'SELECT bio FROM ${student.name}',
                    secondRowSelect: 'SELECT bio FROM ${student.name}',
                    orderByColumn: 'bio',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              ButtonsWidget(
                title: 'Union All',
                onTap: () async {
                  await QueryBuilder.union(
                    type: UnionType.unionAll,
                    firstRowSelect: 'SELECT bio FROM ${student.name}',
                    secondRowSelect: 'SELECT bio FROM ${student.name}',
                    orderByColumn: 'bio',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              // ~ Inner Join
              const Center(child: Text('Join')),
              ButtonsWidget(
                title: 'Inner Join',
                onTap: () async {
                  JoinType.inner;
                  JoinType.left;
                  await QueryBuilder.join(
                    type: JoinType.inner,
                    rowSelect:
                        'SELECT ${collage.name}.address,${student.name}.name FROM ${student.name}',
                    secondTableName: collage.name,
                    firstColumnName: '${collage.name}.collageId',
                    secondColumnName: '${student.name}.collageId',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              // ~ left Join
              ButtonsWidget(
                title: 'Left Join',
                onTap: () async {
                  await QueryBuilder.join(
                    type: JoinType.left,
                    rowSelect:
                        'SELECT ${collage.name}.address,${student.name}.name FROM ${student.name}',
                    secondTableName: collage.name,
                    firstColumnName: '${collage.name}.collageId',
                    secondColumnName: '${student.name}.collageId',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
              // ~ Row Query

              const Center(child: Text('Row Query')),

              ButtonsWidget(
                title: 'Row Query',
                onTap: () async {
                  await QueryBuilder.rawQuery(
                    'query',
                    whenError: (error) {
                      log('Error 👾: $error');
                    },
                  );
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class ButtonsWidget extends StatelessWidget {
  const ButtonsWidget({
    super.key,
    required this.onTap,
    required this.title,
  });
  final VoidCallback onTap;
  final String title;
  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: ElevatedButton(
        onPressed: onTap,
        child: Text(title),
      ),
    );
  }
}

// ~ dummy data

TableModel student = TableModel(
  name: 'Student',
  columns: [
    ColumnModel(
      name: 'studentId',
      isAutoincrement: true,
      isPrimaryKey: true,
      isNotNull: true,
      isInteger: true,
    ),
    ColumnModel(name: 'name', isText: true),
    ColumnModel(name: 'collageId', isInteger: true),
  ],
);

TableModel collage = TableModel(
  name: 'Collage',
  columns: [
    ColumnModel(
      name: 'collageId',
      isAutoincrement: true,
      isPrimaryKey: true,
      isNotNull: true,
      isInteger: true,
    ),
    ColumnModel(name: 'name', isVarchar: true, varcharCharCount: 50),
    ColumnModel(name: 'address', isText: true),
  ],
);
